rasa/engine/training/graph_trainer.py

```diff
@@ -0,0 +1,226 @@
+            model_storage=self._model_storage,
+            execution_context=ExecutionContext(graph_schema=pruned_training_schema),
+            hooks=hooks,
+        )
+
+        logger.debug("Running the pruned train graph with real node execution.")
+
+        graph_runner.run()
+
+        domain = Domain.from_path(domain_path)
+        model_metadata = self._model_storage.create_model_package(
+            output_filename, train_schema, predict_schema, domain
+        )
+
+        return self._graph_runner_class.create(
+            graph_schema=predict_schema,
+            model_storage=self._model_storage,
+            execution_context=ExecutionContext(
+                graph_schema=predict_schema, model_id=model_metadata.model_id
+            ),
+        )
+
+    def _fingerprint_and_prune(self, train_schema: GraphSchema) -> GraphSchema:
+        """Runs the graph using fingerprints to determine which nodes need to re-run.
+
+        Nodes which have a matching fingerprint key in the cache can either be removed
+        entirely from the graph, or replaced with a cached value if their output is
+        needed by descendent nodes.
+
+        Args:
+            train_schema: The train graph schema that will be run in fingerprint mode.
+
+        Returns:
+            A new, potentially smaller and/or cached, graph schema.
+        """
+        fingerprint_schema = self._create_fingerprint_schema(train_schema)
+
+        fingerprint_graph_runner = self._graph_runner_class.create(
+            graph_schema=fingerprint_schema,
+            model_storage=self._model_storage,
+            execution_context=ExecutionContext(graph_schema=fingerprint_schema),
+        )
+
+        logger.debug("Running the train graph in fingerprint mode.")
+        fingerprint_run_outputs = fingerprint_graph_runner.run()
+
+        pruned_training_schema = self._prune_schema(
+            train_schema, fingerprint_run_outputs
+        )
+        return pruned_training_schema
+
+    def _create_fingerprint_schema(self, train_schema: GraphSchema) -> GraphSchema:
+        fingerprint_schema = copy.deepcopy(train_schema)
+        for node_name, schema_node in fingerprint_schema.nodes.items():
+            # We make every node a target so that `graph_runner.run(...)` returns
+            # the output for each node. We need the output of each node
+            # to decide which nodes we can prune.
+            schema_node.is_target = True
+
+            # We do not replace the input nodes as we need an up-to-date fingerprint of
+            # any input data to the graph. This means we can prune according to what
+            # has actually changed.
+            if not schema_node.is_input:
+                FingerprintComponent.replace_schema_node(schema_node, self._cache)
+        return fingerprint_schema
+
+    def _prune_schema(
+        self,
+        schema: GraphSchema,
+        fingerprint_run_outputs: Dict[Text, Union[FingerprintStatus, Any]],
```

Comment:
.... Isn't the `Any` always a `Cachable` here?

Reply:
Nope. You could say that `Any` is always a `Fingerprintable` though

Reply:
Hm, but `CachedComponent` has a `Cachable` `self._output` and `replace_schema_node` configures the class to be a `CachedComponent`  and the config to have output set to something that should be.... `Cachable` then?

Reply:
`_prune_schema` operates on the entire graph after the fingerprint run. During the fingerprint run we run `FingerprintComponent` (which return `FingerprintStatus` objects) **and** input nodes (which return `Any`/ `Fingerprintable`). Input nodes usually load files and return something which not necessarily has to be cacheable. Only `fingerprintable`. after `_prune_schema` we either only have `CachedComponents` with a `Cacheable` output in the graph or a mix of input, `Cacheable` and other nodes which need running.

Reply:
That workflow makes sense to me. What I'm struggling with is that `self._output` of a `CachedComponent` should always be `Cacheable` but it is ok that `replace_schema_node(cls, node: SchemaNode, output: Any)` uses `Any` (and what `replace_schema_node` does is just describe how a `CachedComponent` should be initialised... with an output of type `Any` and not something `Cacheable`). .... So should the type annotations in `CachedComponent` just be `Any` or am I missing something here?

Reply:
You are right, that is inconsistent. The thing is, we use CachedComponent for things where the output was never actually in the cache (see `rasa/engine/training/graph_trainer.py:221`).
So actually `CachedComponent._output` should just be `Any`. 
It doesn't really care if the value came from a cache or not. I guess the component could also have been called something like `ConstantValueComponent`.

Reply:
Since the tests run with non-cacheable things there should also be nothing downstream that expects them to have "from_cache" / where that output might have to be loaded from disk? - We could also call it something like `OutputInMemoryComponent` (doesn't look nice but maybe closer to what it does? 🤔 )

Reply:
Or `PrecomputedValueProvider` (as we call all the other things `*Provider` as well e.g. `TrainingDataProvider` or `DomainProvider`). Or `InmemoryProvider`?

Reply:
@ka-bu @joejuzl I think this is the last open question 🎉

Reply:
Personally I would probably just leave it as is - as I think in most situations it's the most understandable name. However I don't feel super strongly.

Reply:
@ka-bu What do you think? I feel the same as Joe (we can also always rename it somewhen later)

Reply:
I'd prefer one of the Cache-free *provider names cause my brain is strongly like "caching = fingerprinting+store"

Reply:
But we only use these after the fingerprint run, i.e. in a cached run. It's just that it can be wrapping the output of an input node directly to stop running the input nodes twice.

Reply:
went with `PrecomputedValueProvider`

Reply:
That's true but I think the new name still makes it purpose very clear.